metadata/metadata.go

```diff
@@ -105,3 +149,53 @@ func safeCopy(destination map[string]string, key, val string) {
 		destination[key] = val
 	}
 }
+
+// CopyAnnotationsWithPrefixReplacement copies all annotations from a source object to a destination object where the key matches the specified sourcePrefix.
+// If destinationPrefix is different from sourcePrefix, the sourcePrefix will be replaced while performing the copy.
+func CopyAnnotationsWithPrefixReplacement(source, destination v1.Object, sourcePrefix, destinationPrefix string) error {
+	if _, err := GetAnnotationsWithPrefix(source, sourcePrefix); err != nil {
+		return err
+	}
+	if destination.GetAnnotations() == nil {
+		destination.SetAnnotations(make(map[string]string))
+	}
+	copyWithPrefixReplacement(source.GetAnnotations(), destination.GetAnnotations(), sourcePrefix, destinationPrefix)
+	return nil
+}
+
+// CopyAnnotationsByPrefix copies all annotations from a source object to a destination object where the key matches the specified sourcePrefix.
+func CopyAnnotationsByPrefix(source, destination v1.Object, prefix string) error {
+	if _, err := GetAnnotationsWithPrefix(source, prefix); err != nil {
+		return err
+	}
+	if destination.GetAnnotations() == nil {
+		destination.SetAnnotations(make(map[string]string))
+	}
+	copyByPrefix(source.GetAnnotations(), destination.GetAnnotations(), prefix)
+	return nil
+}
+
+// CopyLabelsWithPrefixReplacement copies all labels from a source object to a destination object where the key matches the specified sourcePrefix.
+// If destinationPrefix is different from sourcePrefix, the sourcePrefix will be replaced while performing the copy.
+func CopyLabelsWithPrefixReplacement(source, destination v1.Object, sourcePrefix, destinationPrefix string) error {
+	if _, err := GetLabelsWithPrefix(source, sourcePrefix); err != nil {
+		return err
+	}
+	if destination.GetLabels() == nil {
+		destination.SetLabels(make(map[string]string))
+	}
```

Comment:
I don't think you need the lines above, isn't it? `copyWithPrefixReplacement` already filter by prefix and makes the copy.

The check in here should be something like:
```
if source == nil || destination == nil {
    return errors.New("objects cannot be nil")
}
```

Reply:
The check and code in line#184-186 code is to avoid the error when the Labels of  the destination object is empty.  The code you suggested is to avoid the destination object is nil. They are different. So, I updated the code and kept them both.

Reply:
@davidmogar , can you please help to review the PR again? Thanks

Reply:
That cannot happen. The nil value for maps is an empty map. I have an example here: https://go.dev/play/p/_3wZ0gaDAoj

Reply:
Code should be something like this:
```
func CopyLabelsByPrefix(source, destination v1.Object, prefix string) error {
	if source == nil || destination == nil {
		return errors.New("object cannot be nil")
	}

	copyByPrefix(source.GetLabels(), destination.GetLabels(), prefix)
	return nil
}
```

If you could keep that empty line so it's more readable it would be great!

Reply:
> That cannot happen. The nil value for maps is an empty map. I have an example here: https://go.dev/play/p/_3wZ0gaDAoj

The nil value for destination map will not lead to error.  But the labels should be copied to the destination when the map of destination is nil. Is that make sense to you?

Reply:
After discussing offline, Jinq is right about having to initialize the destination. So the code should be:
```
func copyWithPrefixReplacement(source, destination map[string]string, sourcePrefix, destinationPrefix string) {
        if destination == nil {
		destination = map[string]string{}
	}

	for key, value := range source {
		if strings.HasPrefix(key, sourcePrefix) {
			newKey := key
			if sourcePrefix != destinationPrefix {
				newKey = strings.Replace(key, sourcePrefix, destinationPrefix, 1)
			}
			destination[newKey] = value
		}
	}
}
```

Reply:
Thanks. But with above change,  the test case can't pass  - It("should copy all the labels matching the given prefix when Labels in destination is empty", func() {
podSource := &corev1.Pod{
ObjectMeta: v1.ObjectMeta{
Labels: map[string]string{"foo": "bar", "baz": "qux"},
},
}
podDest := &corev1.Pod{
ObjectMeta: v1.ObjectMeta{},
}
CopyLabelsWithPrefixReplacement(podSource, podDest, "fo", "ba")
Expect(podDest.Labels).To(gstruct.MatchAllKeys(gstruct.Keys{
"bao": Equal("bar"),
}))
})
So, the space allocation for the Labels is put [here] (https://github.com/redhat-appstudio/operator-toolkit/blob/b4e94cd47d6782c8ba9efd9687a04023a47c9378/metadata/metadata.go#L112C2-L114C3), and other similar places.